frontend/src/main/scala/bloop/dap/DebugSession.scala

```diff
@@ -0,0 +1,182 @@
+  private val runningDebuggee = Atomic(CancelableFuture.unit)
+
+  /**
+   * Schedules the start of the debugging session.
+   *
+   * For a session to start, two executions must happen independently in a
+   * non-blocking way: the debuggee process is started in the background and
+   * the DAP server starts listening to client requests in an IO thread.
+   */
+  def startDebuggeeAndServer(): Unit = {
+    ioScheduler.executeAsync(() => {
+      if (isStarted.compareAndSet(false, true)) {
+        val logger =
+          new DebugSessionLogger(this, address => debugAddress.success(address), initialLogger)
+        val debuggeeHandle = startDebuggee(logger).runAsync(ioScheduler)
+        isCancelled.getAndTransform { isCancelled =>
+          if (isCancelled) debuggeeHandle.cancel()
+          else runningDebuggee.set(debuggeeHandle)
+          isCancelled
+        }
+
+        try super.run()
+        finally {
+          exitStatusPromise.trySuccess(DebugSession.Terminated); ()
+        }
+      }
+    })
+  }
+
+  override def dispatchRequest(request: Request): Unit = {
+    val requestId = request.seq
+    request.command match {
+      case "launch" =>
+        launchedRequests.add(requestId)
+        val _ = Task
+          .fromFuture(debugAddress.future)
+          .map(DebugSession.toAttachRequest(requestId, _))
+          .foreachL(super.dispatchRequest)
+          .runAsync(ioScheduler)
+
+      case "disconnect" =>
+        if (DebugSession.shouldRestart(request))
+          exitStatusPromise.trySuccess(DebugSession.Restarted)
+        runningDebuggee.get.cancel()
+        super.dispatchRequest(request)
+
+      case _ => super.dispatchRequest(request)
+    }
+  }
+
+  override def sendResponse(response: Response): Unit = {
+    val requestId = response.request_seq
+    response.command match {
+      case "attach" if launchedRequests(requestId) =>
+        // Trick dap4j into thinking we're processing a launch instead of attach
+        response.command = Command.LAUNCH.getName
+        super.sendResponse(response)
+      case "disconnect" if requestId == DebugSession.InternalRequestId =>
+        // Request sent by the session itself, don't return to client
+        socket.close()
+      case _ => super.sendResponse(response)
+    }
+  }
+
+  override def sendEvent(event: Events.DebugEvent): Unit = {
+    if (event.`type` == "exited") {
+      debuggeeExited.success(())
+    }
+
+    super.sendEvent(event)
+  }
+
+  def exitStatus: Task[DebugSession.ExitStatus] = {
+    Task.fromFuture(exitStatusPromise.future)
+  }
+
+  /**
+   * Cancels the background debuggee process, the DAP server and closes the socket.
+   */
+  def cancel(): Unit = {
+    if (isCancelled.compareAndSet(false, true)) {
+      runningDebuggee.get.cancel()
+      Task
+        .fromFuture(debuggeeExited.future)
+        .map(_ => DebugSession.disconnectRequest(DebugSession.InternalRequestId))
```

Comment:
I'm down with having the DAP server send all of those exited and terminated events on its own, but then we need to figure out several important aspects that are not clear to me yet. Here's a few questions:
  
* How does cancelling the debuggee affect the DAP server running in the background? If the VM the DAP server is attached to exits because we cancel it, what exception in received by the DAP server and how is it handled?
* Why do we need to send this disconnect request manually? Is the exit of the VM not forcing the delivery of exited and terminated events?
* Should we not try to get the DAP server to disconnect from the VM first and then cancel the debuggee when the exited/terminated events have been sent out? At the point we should also be able to close the socket imediately.
* Why are we also cancelling the debuggee before the cancellation logic runs and inside  the handling of the internal disconnect in `dispatchRequest`?
* Can we get rid of the delay and ensure that we clean up resources at the very moment where we cancel everything so that we can instantiate another debug session (in case of a restart) from a clean slate?
* The handling of `dispatchRequest` in `ProtocolServer` is special casing some VM disconnected errors. If we were to do something special, should we be trying to handle benign VM-related errors and return successful responses to the client (assuming we might not be doing that already)?

Reply:
It looks like this internal disconnect request is indeed redundant. Also, I think I have found a race condition in the IntelliJ UTestRunner (in Intellij it sometimes "hangs" with one test case marked "running" even though it clearly finished (last statement was executed)). I have run the same test suites in sbt and didn't encounter the issues in more than 30 runs (Intellij usually failed 1 out of 10 runs).

To answer your questions:
>  If the VM (...) exits because we cancel it (...) how is it handled?

If debuggee exits unexpectedly it emits VMDeath and VMDisconnected events. Those are handled by sending the Exited and Terminated events respectively.

> Why do we need to send this disconnect request manually?

It turns out we don't have to. There is some cleanup logic in DisconnectHandler but it is used only when the Server is launching the debuggee, not attaching to it like in our case. 

> Can we get rid of the delay and ensure that we clean up resources at the very moment where we cancel everything so that we can instantiate another debug session (in case of a restart) from a clean slate?

When we cancel the Session there is no restart. We stop the debuggee, client socket and the server.
Also, the timeout is necessary if the debuggee for some reason does not close. Without the timeout we would not close the socket in this case.